docs/3.x/extend/graphql.md

```diff
@@ -0,0 +1,1169 @@
+{
+  widgets {
+    dateCreated @formatDateTime (format: "Y-m-d")
+  }
+}
+```
+
+Craft’s included directives apply exclusively to requested fields, though they may be applied in mutations and numerous parts of the type system.
+
+A directive needs to provide a name, description, and the relevant query location(s) it can be applied. It can optionally specify [arguments](#arguments).
+
+### Example Directive Class
+
+This simple example returns the field value, replacing any instances of `foo` with `bar`.
+
+```php
+namespace mynamespace\gql\directives;
+
+use craft\gql\base\Directive;
+use craft\gql\GqlEntityRegistry;
+use GraphQL\Type\Definition\Directive as GqlDirective;
+
+class BarTheFoo extends Directive
+{
+
+    public static function create(): GqlDirective
+    {
+        if ($type = GqlEntityRegistry::getEntity(self::name())) {
+            return $type;
+        }
+
+        $type = GqlEntityRegistry::createEntity(static::name(), new self([
+            'name' => static::name(),
+            'locations' => [
+                DirectiveLocation::FIELD,
+            ],
+            'description' => 'Replace `foo` with `bar`.',
+            'args' => [],
+        ]));
+
+        return $type;
+    }
+
+    public static function name(): string
+    {
+        return 'fooTheBar';
+    }
+
+    public static function apply($source, $value, array $arguments, ResolveInfo $resolveInfo)
+    {
+        return str_replace('foo', 'bar', (string)$value);
+    }
+}
+```
+
+### Registering Directives
+
+You can register your directive by appending its class name to the `directives` array on the [registerGqlDirectives](craft3:craft\services\Gql::EVENT_REGISTER_GQL_DIRECTIVES) event object:
+
+```php
+use mynamespace\gql\directives\BarTheFoo;
+use craft\events\RegisterGqlDirectivesEvent;
+use craft\services\Gql;
+use yii\base\Event;
+
+Event::on(
+    Gql::class,
+    Gql::EVENT_REGISTER_GQL_DIRECTIVES,
+    function(RegisterGqlDirectivesEvent $event) {
+        $event->directives[] = BarTheFoo::class;
+    }
+);
+```
+
+## Mutations
+
+A Mutation class defines named mutations that should be available, including consideration for scope access. Each mutation can have its own [arguments](#arguments) and type, and it will need a mutation resolver for modifying data using Craft’s API.
+
+Our example could provide a `saveWidget` mutation for saving a new widget:
+
+::: code
+```graphql{2} GraphQL Query
+mutation NewWidget($title: String) {
+  createWidget(title: $title) {
+    id
+    title
+  }
+}
+
+# query variables:
+# {
+#  "title": "My Glorious Mutated Widget",
+# }
+```
+```json JSON Response
+{
+  "data": {
+    "createWidget": {
+        "id": "123",
+        "title": "My Glorious Mutated Widget"
+    }
+}
+```
+:::
+
+::: tip
+Like the name-getting methods mentioned earlier, Craft’s base elements include a <craft3:craft\base\Element::gqlMutationNameByContext()> method for providing context-sensitive type names.
+:::
+
+### Input Types
```

Comment:
Worth mentioning the value normalizers, that input types can add (https://github.com/craftcms/cms/blob/develop/src/gql/types/input/Matrix.php#L95-L138)

This is a step that allows the input type to re-arrange (or do anything, really), before the data is passed on to saving. For example, Matrix input type uses this to tweak the data structure from "what's possible in GQL" to "What a Matrix field expects the data to be".

I also know @timkelty has used this with success to index an Asset by it's URI and swap out the URI with the Asset ID, so that a relationship gets saved with an Asset that didn't exist prior the mutation.

Reply:
Indeed! Here's the thread where we hashed that out: https://github.com/craftcms/cms/issues/6398

Reply:
Thank you both!

If I’m following along right, this can be used to open a little window between receiving valid GraphQL mutation data and handing it off to the mutation resolver for Craft to save stuff.

1. Craft is the one supporting this `'normalizeValue'` config option, which otherwise has no meaning for `webonyx/graphql-php`.
2. If you provide that `'normalizeValue'` key, its config value needs to be a function that can mess with whatever field `$value` is on its way to the resolver.
3. @timkelty’s [final solution](https://github.com/craftcms/cms/issues/6398#issuecomment-661136684) wound up modifying the expected type for the `emailAttachments` mutation field, then sneaking in _after the entry’s populated but before it’s saved_ to swap in Craft-appropriate `authorId` and `emailAttachments` values.

For item 3 (ignoring `authorId`) Tim’s answer _could have_ listened for `EVENT_REGISTER_GQL_MUTATIONS` and alternatively replaced `$new['args']['emailAttachments'] = Type::listOf(Type::string());` with a list of custom input object types that use `'normalizeValue'` to take file path strings and turn them into Asset IDs.

Pseudo code:

```php
use supertim\gql\types\input\AssetPath;

Event::on(
    Gql::class,
    Gql::EVENT_REGISTER_GQL_MUTATIONS,
    function (RegisterGqlMutationsEvent $event) {
        $new = $event->mutations['save_emailOrders_emailOrders_Entry'];
        $new['name'] = 'save_emailOrder';
        $new['args']['emailAttachments'] = Type::listOf(AssetPath::getType());

        $event->mutations[] = $new;
    }
);
```

```php
namespace supertim\gql\types\input;

class AssetPath
{
    public static function getType()
    {
        $typeName = 'AssetPath';
        
        $inputType = GqlEntityRegistry::createEntity($typeName, new InputObjectType([
            'name' => $typeName,
            // not real sure about this part
            'fields' => function() {
                return Type::string();
            },
            'normalizeValue' => function($value) {
                if (is_string($value)) {
                    $volume = Craft::$app->volumes->getVolumeByHandle('orderAttachments');
        
                    $path = trim($value);
                    $pathInVolume = trim(StringHelper::replaceBeginning($path, Craft::parseEnv($volume->subfolder), ''), '/');
                    $asset = Craft::$app->assetIndexer->indexFile($volume, $pathInVolume);
                    $value = $asset->id;
                }
        
                return $value;
            },
        ]));
        
        return $inputType;
    }
}
```

Reply:
This example doesn’t work as written because `fields` is invalid and returning `'type' => Type::string()` instead ceases to make it an input object. Leaving out the example for now until I can come back with a useful illustration.